pyro/contrib/gp/kernels/kernel.py

```diff
@@ -52,3 +68,238 @@ def _slice_input(self, X):
+            self._subkernels[kern.name] = kern
+
+
+def _Horner_evaluate(x, coef):
+    """
+    Evaluates the value of a polynomial accoridng to Horner's method.
+    """
+    # https://en.wikipedia.org/wiki/Horner%27s_method
+    n = len(coef) - 1
+    b = coef[n]
+    for i in range(n-1, -1, -1):
+        b = coef[i] + b * x
+    return b
+
+
+class Warping(Deriving):
+    """
+    Creates a new kernel according to ``k_new(x, z) = q(k(f(x), f(z)))``,
+    where ``f := iwarping_fn`` is a input warping function, and ``q`` is a polynomial
+    with non-negative coefficients ``owarping_coef``.
+
+    :param iwarping_fn: The input warping function, must be callable.
+    :param list owarping_coef: A list of coefficients of the output warping polynomial.
+        These coefficients must be non-negative.
+    """
+
+    def __init__(self, kern, iwarping_fn, owarping_coef, name=None):
+        super(Warping, self).__init__(kern, name)
+
+        self.iwarping_fn = iwarping_fn
+
+        if owarping_coef is not None:
+            for coef in owarping_coef:
+                if not isinstance(coef, int) and coef < 0:
+                    raise ValueError("Coefficients of the polynomial must be a non-negative integer.")
+            if len(owarping_coef) < 2 and sum(owarping_coef) == 0:
+                raise ValueError("The ouput warping polynomial should have a degree of at least 1.")
+        self.owarping_coef = owarping_coef
+
+    def forward(self, X, Z=None, diag=False):
+        if self.iwarping_fn is None:
+            K_iwarp = self.kern(X, Z, diag)
+        elif Z is None:
+            K_iwarp = self.kern(self.iwarping_fn(X), None, diag)
+        else:
+            K_iwarp = self.kern(self.iwarping_fn(X), self.iwarping_fn(Z), diag)
+
+        if self.owarping_coef is None:
+            return K_iwarp
+        else:
+            return _Horner_evaluate(K_iwarp, self.owarping_coef)
+
+
+class Exponent(Deriving):
+    """
+    Creates a new kernel according to ``k_new(x, z) = exp(k(x, z))``.
+    """
+
+    def forward(self, X, Z=None, diag=False):
+        return self.kern(X, Z, diag).exp()
+
+
+class VerticalScaling(Deriving):
+    """
+    Creates a new kernel according to ``k_new(x, z) = f(x)k(x, z)f(z)``,
+    where ``f := vscaling_fn`` is a vertical scaling function.
+
+    :param vscaling_fn: The vertical scaling function, must be callable.
+    """
+
+    def __init__(self, kern, vscaling_fn, name=None):
+        super(VerticalScaling, self).__init__(kern, name)
+
+        self.vscaling_fn = vscaling_fn
+
+    def forward(self, X, Z=None, diag=False):
+        if diag:
+            return self.vscaling_fn(X) * self.kern(X, Z, diag) * self.vscaling_fn(X)
+        elif Z is None:
+            vscaled_X = self.vscaling_fn(X).unsqueeze(1)
+            return vscaled_X * self.kern(X, Z, diag) * vscaled_X.t()
+        else:
+            return self.vscaling_fn(X).unsqueeze(1) * self.kern(X, Z, diag) * self.vscaling_fn(Z).unsqueeze(0)
```

Comment:
Will we ever support batching? If so we should index from the right: `...unsqueeze(-1) * ...unsqueeze(-2)`.

Reply:
@fritzo I think that I need to find some references to GP with batch. Will address this issue in a future pull request, when I figure out how batching works. Right now I think that it is good to leave as it is, so the tests for GP with batch will throw wrong answers.